src/cpp/src/threadpool.hpp

```diff
@@ -0,0 +1,70 @@
+#include <condition_variable>
+#include <functional>
+#include <future>
+#include <iostream>
+#include <mutex>
+#include <queue>
+#include <thread>
+#include <utility>
+#include <atomic>
+
+class ThreadPool {
+
+private:
+    std::vector<std::thread> threads;
+    std::queue<std::function<void()>> tasks;
+    std::mutex queue_mutex;
+    std::condition_variable cv;
+    std::atomic<bool> stop{false};
+
+public:
+    ThreadPool(const ThreadPool& rhs) = delete;
+    ThreadPool(ThreadPool&& rhs) = delete;
+    ThreadPool(size_t num_threads = std::thread::hardware_concurrency())
```

Comment:
BTW, maybe we can expose num_threads in Sampler ctor? It will 1 by default, and `std::thread::hardware_concurrency()` for CB
how do you think?

Reply:
Makes sense. Maybe even add this as a parameter to CB pipeline constructor or some public method? For someone using offline application with one request at a time parallel sampling will not be much of an improvement and for high concurrency someone could specify it via CB pipeline constructor parameter that would be passed to the sampler.
Perhaps it would be useful to allow user decide?